### PR TITLE
Throw on unresolved options

### PR DESCRIPTION
**Pull request by oldergod**
This is what Protoc does.
I think we should just turn it on without going through an option.
It might break some schema but as we can see in the files I had to fix, something like `redacted` could be NOT applied without nobody noticing it. It actually happened a few days ago internally.

### CODE REVIEW

File: wire-library/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Options.kt [hunk 1]
```diff
@@ -96,6 +96,7 @@ class Options(
         path = resolveFieldPath(packageName + "." + option.name, extensionsForType.keys)
       }
       if (path == null) {
+        linker.addError("unable to resolve ${option.name}")
```
Comment (JakeWharton) [inline wire-library/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Options.kt:99]:
```suggestion
        linker.addError("unable to resolve option '${option.name}'")
```

Comment (oldergod) [reply wire-library/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Options.kt]: Fixed

**Comment by oldergod** [comment]
Added rule for file options

File: wire-library/wire-compiler/src/test/java/com/squareup/wire/WireCompilerErrorTest.kt [hunk 2]
```diff
@@ -140,6 +140,7 @@ class WireCompilerErrorTest {
           |multiple enums share constant QUIX:
           |  1. com.squareup.protos.test.Bar.QUIX (/source/test.proto:4:3)
           |  2. com.squareup.protos.test.Bar2.QUIX (/source/test.proto:10:3)
+          |  for file /source/test.proto
```
Comment (oldergod) [inline wire-library/wire-compiler/src/test/java/com/squareup/wire/WireCompilerErrorTest.kt:143]: That's unfortunate.
I cannot come up with a fix for this specific case, and I also think it's not a big of a deal with we add this line for this error?

Comment (JakeWharton) [reply wire-library/wire-compiler/src/test/java/com/squareup/wire/WireCompilerErrorTest.kt:143]: Seems fine

File: wire-library/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Linker.kt [hunk 3]
```diff
@@ -449,6 +449,10 @@ class Linker {
     val error = StringBuilder()
     error.append(message)
 
+    val contextStack = this.contextStack.toMutableList()
+    if (contextStack.size > 1) {
+      contextStack.removeAll { it is ProtoFile }
```
Comment (swankjesse) [inline wire-library/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Linker.kt:454]: perfect

Comment (swankjesse) [reply wire-library/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Linker.kt:454]: Screen Shot 2020-07-17 at 21 52 56

I spoke too soon! Needs to check if any is not a ProtoFile.
```
    if (contextStack.any { it !is ProtoType }) {
      contextStack.removeAll { it is ProtoFile }
    }
```

Comment (oldergod) [reply wire-library/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Linker.kt:454]: Fixed, thanks for checking!

File: wire-library/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Linker.kt [hunk 4]
```diff
@@ -94,14 +94,14 @@ class Linker {
...
-      fileLinker.linkOptions(syntaxRules)
+      fileLinker.linkOptions(syntaxRules, validate = true)
     }
 
     // For compactness we'd prefer to link the options of source files only. But we link file
     // options on referenced files to make sure that java_package is populated.
     while (fileOptionsQueue.isNotEmpty()) {
       val fileLinker = fileOptionsQueue.poll()!!
-      fileLinker.requireFileOptionsLinked()
+      fileLinker.requireFileOptionsLinked(validate = false)
```
Comment (swankjesse) [inline wire-library/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Linker.kt:104]: NICE

**Comment by swankjesse** [comment]
Confirmed this works with our internal protos.

**Comment by oldergod** [comment]
Thank you!

**Pull request merged**